fud2/fud-core/src/exec/plan/egg_planner.rs

```diff
@@ -0,0 +1,229 @@
+use super::{
+    super::{OpRef, Operation, StateRef},
+    FindPlan, Step,
+};
+use cranelift_entity::PrimaryMap;
+use egg::{define_language, rewrite, Id, Pattern, RecExpr, Rewrite, Runner};
+
+#[derive(Debug, Default)]
+pub struct EggPlanner;
+
+define_language! {
+    /// A language to represent a collection of states.
+    /// For example, if there are 3 states and 2 ops, a term `(root (states x x x) (ops x x))`
+    /// represents the absense of all states and all ops. A term `(root (states 3 x x) (ops x 2))`
+    /// represents a a pair of sets, the first containing a single state with `StateRef` 3 and the
+    /// second containing a single op with `OpRef` 2.
+    enum StateLanguage {
+        // The root of a term. This is used to store a pair of "states" and "ops".
+        "root" = Root([Id; 2]),
+        // A list of states.
+        "states" = States(Box<[Id]>),
+        // A list of ops.
+        "ops" = Ops(Box<[Id]>),
+        // Symbolizes the absense of a state or op.
+        "xxx" = X,
+        // A ref, refering to either a StateRef or OpRef depending on context.
+        Ref(u32),
+    }
+}
+
+/// Construct string which can be parsed to a `root` term in `StateLanguage`.
+///
+/// `states` are in included states in the term, and `through` are the included ops. If
+/// `states_use_x` is set, then all absent states will be marked with `x`. If it is not set, these
+/// unused states will instead be marked with a unique variable to create a string parsable into a
+/// pattern. `ops_use_x` is similar but applies to unused ops.
+///
+/// `all_states` is an ordered list of all states. `all_ops` is an ordered list of all ops.
+fn language_string(
```

Comment:
It doesn't look to hard. It does increase code dup a little (one function call gets replaced by almost the call's contents) because, as you note, patterns have to be specified separately. 

I do agree though, seems like an overall win. Done!